pkg/event-consumers/kafka/kafka-consumer.go

```diff
@@ -0,0 +1,155 @@
+package kafka
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"github.com/bsm/sarama-cluster"
+	"github.com/kubeless/kubeless/pkg/utils"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+var (
+	stopM     map[string](chan struct{})
+	stoppedM  map[string](chan struct{})
+	consumerM map[string]bool
+	brokers   string
+)
+
+func init() {
+	stopM = make(map[string](chan struct{}))
+	stoppedM = make(map[string](chan struct{}))
+	consumerM = make(map[string]bool)
+
+	// taking brokers from env var
+	brokers = os.Getenv("KAFKA_BROKERS")
+	if brokers == "" {
+		brokers = "kafka.kubeless:9092"
+	}
+}
+
+// createConsumerProcess gets messages to a Kafka topic from the broker and send the payload to function service
+func createConsumerProcess(broker, topic, funcName, ns, consumerGroupID string, stopchan, stoppedchan chan struct{}) {
+	// Init config
+	config := cluster.NewConfig()
+
+	config.Consumer.Return.Errors = true
+	config.Group.Return.Notifications = true
+	config.Consumer.Offsets.Initial = sarama.OffsetNewest
+
+	// Init consumer
+	brokersSlice := []string{broker}
+	topicsSlice := []string{topic}
+
+	consumer, err := cluster.NewConsumer(brokersSlice, consumerGroupID, topicsSlice, config)
+	if err != nil {
+		logrus.Fatalf("Failed to start consumer: %v", err)
+	}
+	defer consumer.Close()
+
+	logrus.Infof("Started Kakfa consumer Broker: %v, Topic: %v, Function: %v, consumerID: %v", broker, topic, funcName, consumerGroupID)
+
+	// Consume messages, wait for signal to stopchan to exit
+	defer close(stoppedchan)
+	for {
+		select {
+		case msg, more := <-consumer.Messages():
+			if more {
+				logrus.Infof("Received Kafka message Partition: %d Offset: %d Key: %s Value: %s ", msg.Partition, msg.Offset, string(msg.Key), string(msg.Value))
+				//forward msg to function
+				clientset := utils.GetClient()
+				err = sendMessage(clientset, funcName, ns, string(msg.Value))
+				if err != nil {
+					logrus.Errorf("Failed to send message to function: %v", err)
+				}
+				consumer.MarkOffset(msg, "")
```

Comment:
for the moment I would mimic the behavior of a HTTP trigger so yes, I would leave it as "at-most-once" and don't retry in case of error

Reply:
That's fine, but we should document that somewhere. I think people in serverless-land often use queues (instead of eg webhooks) precisely because they _don't_ want to lose events - so we should let them know what they can expect from a kubeless queue.